ApplicationLibCode/ModelVisualization/RivSimWellPipesPartMgr.cpp

```diff
@@ -280,93 +289,238 @@ void RivSimWellPipesPartMgr::buildWellPipeParts( const caf::DisplayCoordTransfor
+
+    const auto characteristicCellSize = eclipseView->ownerCase()->characteristicCellSize();
+    const auto coords                 = pbd.m_pipeGeomGenerator->pipeCenterCoords();
+
+    std::set<std::pair<size_t, size_t>> resultPointWithValve;
+    for ( size_t i = 0; i < pbd.m_cellIds.size(); i++ )
+    {
+        const auto resultPoint = pbd.m_cellIds[i];
+
+        if ( resultPointWithValve.contains( { resultPoint.gridIndex(), resultPoint.cellIndex() } ) ) continue;
+
+        const auto   segmentStartCoord = coords->get( i );
+        const auto   segmentEndCoord   = coords->get( i + 1 );
+        const auto   segmentLength     = ( segmentEndCoord - segmentStartCoord ).length();
+        const double valveLength       = characteristicCellSize * 0.2;
+
+        // Insert valve geometry if the length of the segment is sufficient
+        if ( resultPoint.isConnectedToValve() && ( valveLength * 2 < segmentLength ) )
+        {
+            resultPointWithValve.insert( { resultPoint.gridIndex(), resultPoint.cellIndex() } );
+
+            const auto segmentDirection = ( segmentEndCoord - segmentStartCoord ).getNormalized();
+
+            // A segment ends at the center of a simulation cell
+            auto valveCenterCoord = ( segmentEndCoord + segmentStartCoord ) / 2.0;
+
+            const cvf::Color3f valveColor                      = cvf::Color3f::ORANGE;
+            const auto         measuredDepthsRelativeToStartMD = { 0.0, 1.0, valveLength - 1.0, valveLength };
+
+            const auto outerValveRadius = wellPathRadius * 1.3;
+            const auto radii            = { wellPathRadius, outerValveRadius, outerValveRadius, wellPathRadius };
+
+            // The location of the valve is adjusted to locate the valve at the center of the segment
+            double locationAdjustment = -valveLength / 2.0;
+
+            std::vector<cvf::Vec3d> displayCoords;
+            for ( const auto& mdRelativeToStart : measuredDepthsRelativeToStartMD )
+            {
+                displayCoords.push_back( valveCenterCoord + ( locationAdjustment + mdRelativeToStart ) * segmentDirection );
+            }
+
+            RivPipeGeometryGenerator::tubeWithCenterLinePartsAndVariableWidth( &pbd.m_valveParts, displayCoords, radii, valveColor );
+
+            auto computeRotationAxisAndAngle = []( const cvf::Vec3f& direction ) {
+                // Compute upwards normal based on direction
+                // Compute the rotation axis and angle between up vector and Z_AXIS
+
+                cvf::Vec3f crossBetweenZAndDirection;
+                crossBetweenZAndDirection.cross( cvf::Vec3f::Z_AXIS, cvf::Vec3f( direction ) );
+
+                cvf::Vec3f upVector;
+                upVector.cross( cvf::Vec3f( direction ), crossBetweenZAndDirection );
+                if ( upVector.z() < 0.0 ) upVector = -upVector;
+
+                cvf::Vec3f rotationAxis;
+                rotationAxis.cross( upVector, cvf::Vec3f::Z_AXIS );
```

Comment:
Is `crossBetweenZAndDirection` equal `rotationAxis`? 

- `crossBetweenZAndDirection` is vector perpendictular to z-axis and segment direction
- `upVector` is the vector perpendicular to the segment direction vector
- `rotationAxis` is the vector perpendicular to the `upVector` and z-axis, which I think can be equal to the first bullet point?

Can be worth testing if one cross product is enough to determine rotation axis

Reply:
I see your point, but testing gives wrong orientation of valves. Original solution kept.